### ISSUE TITLE
Homebrew hangs when installing or updating homebrew/nginx/nginx-full on Yosemite 10.10.5

### ISSUE THREAD

**Issue by dunn** [Homebrew/homebrew-nginx#167]
_From @dustinwilson on March 23, 2016 21:58_

Homebrew hangs after completing nginx's makefile on Yosemite 10.10.5. I've tried this on two different computers, and one had a clean install of homebrew. Nginx finishes making and installing. Everything's within its cellar folder. I also know everything finished because if I copy that folder out while it is hanging, ctrl+C, copy the folder back in, and then `brew link nginx-full` it works just fine. I've also tested on El Capitan, and things work just fine there.

https://gist.github.com/anonymous/0a676790862f861c119d

_Copied from original issue: Homebrew/homebrew#50369_

**Comment by clounie** [comment]
Had this happen on 10.11.3 and 10.11.4 as well. Computer was a Mid 2014 MBP Retina w/3Ghz i7.

However, on my Late 2013 MBP Retina (2.6Ghz i7), it installed fine for 10.11.3.

It always hung after doing 
```
test -d '/usr/local/Cellar/nginx-full/1.8.1/html' 		|| cp -R html '/usr/local/Cellar/nginx-full/1.8.1'
test -d '/usr/local/var/log/nginx' || 		mkdir -p '/usr/local/var/log/nginx'
```

It appears to go past that point (ruby/sh kept on trading off as the active task in the terminal tab), but he let it run overnight and it still never finished.

**Comment by terrybleger** [comment]
Yes, I had the same problem here. Hang with both stable and devel version.

```
brew tap homebrew/homebrew-nginx
brew install nginx-full
```

```
CPU: 8-core 64-bit haswell
OS X: 10.11.4-x86_64
Xcode: 7.3
CLT: 7.3.0.0.1.1457485338
Clang: 7.3 build 703
System Ruby: 2.0.0-p648
```

**Comment by denji** [comment]
Can check, hotfix in case of the simultaneous installation of `homebrew/nginx` and `nginx/nginx-full`.

**Comment by terrybleger** [comment]
Download the source, then

```
./configure \
--prefix=/usr/local/Cellar/nginx-full/1.9.12 \
--sbin-path=/usr/local/Cellar/nginx-full/1.9.12/bin/nginx \
--with-http_ssl_module \
--with-pcre \
--with-ipv6 \
--with-http_v2_module \
--add-module=/usr/local/share/headers-more-nginx-module \
--add-module=/usr/local/opt/passenger/libexec/src/nginx_module \
--with-cc-opt="-I/usr/local/include -I/usr/local/opt/pcre/include -I/usr/local/opt/openssl/include" \
--with-ld-opt="-L/usr/local/lib -L/usr/local/opt/pcre/lib -L/usr/local/opt/openssl/lib" \
--conf-path=/usr/local/etc/nginx/nginx.conf \
--pid-path=/usr/local/var/run/nginx.pid \
--lock-path=/usr/local/var/run/nginx.lock \
--http-client-body-temp-path=/usr/local/var/run/nginx/client_body_temp \
--http-proxy-temp-path=/usr/local/var/run/nginx/proxy_temp \
--http-fastcgi-temp-path=/usr/local/var/run/nginx/fastcgi_temp \
--http-uwsgi-temp-path=/usr/local/var/run/nginx/uwsgi_temp \
--http-scgi-temp-path=/usr/local/var/run/nginx/scgi_temp \
--http-log-path=/usr/local/var/log/nginx/access.log \
--error-log-path=/usr/local/var/log/nginx/error.log

make && make install

brew link nginx-full
```

Installed successfully in less then 2 minutes. I don't know why.

**Comment by julianhirt** [comment]
I have the same issue here. Mac Mini on 10.11.4

**Comment by CarloQ** [comment]
+1

**Comment by aaronbruckner** [comment]
I'm not the only one! I've spent many hours looking into this and had the exact same hang as you CarloQ. I narrowed it down to a failure outside the formula (I believe). I determined this by slowly adding logs all the way from the initial brew start at brew.rb and tracing many levels lower. 


**TLDR** how to reproduce - go to /usr/local/Library/Homebrew/utils.rb line 182. Replace 

```
  def self._system(cmd, *args)
    pid = fork do
      yield if block_given?
      args.collect!(&:to_s)
      exec(cmd, *args) rescue nil
      exit! 1 # never gets here unless exec failed
    end
    Process.wait(pid)
    $?.success?
  end
```
with
```
  def self._system(cmd, *args)
    pid = fork do
      yield if block_given?
      args.collect!(&:to_s)
      exec(cmd, *args) rescue nil
      exit! 1 # never gets here unless exec failed
    end
    puts "Process ID: " + pid.to_s
    Process.wait(pid)
    puts "Process complete!"
    $?.success?
  end
```
This just wraps the wait with console logs. You should see that the line Process.wait(pid) hangs forever.

**To fix** - I HAVE NO IDEA IF THIS WILL TOTALLY SCREW EVERYTHING UP BUT IT WORKED FOR ME... 

Update /user/local/Library/Homebrew/utils.rb line 203 from
```
  def self.git_head
    return unless Utils.git_available?
    HOMEBREW_REPOSITORY.cd { `git rev-parse --verify -q HEAD 2>/dev/null`.chuzzle }
  end
```
TO
```
  def self.git_head
    return
    HOMEBREW_REPOSITORY.cd { `git rev-parse --verify -q HEAD 2>/dev/null`.chuzzle }
  end
```

This avoids the never exiting process and seems (maybe) safe as it conditionally exits and doesn't return head.

**Longer version** Tracing down to this problem.

My brew ruby files are located at /usr/local/Library/Homebrew. The initial starting point for brew commands is brew.rb. Eventually, function install is invoked in build.rb. This successfully invokes the formula's install function (much to my initial surprise).Tab.create (located in tab.rb) is invoked after the formula install completes. Tab.create seems to compile a book keeping record of configurations and timestamps and this is where I notice it never returned. 

One of the arguments Tab.create collets is the git head which is done by calling git_head located in utils.rb. git_head then calls the function git_available located in git.rb. Git_available attempts to execute "HOMEBREW_ENV_PATH/scm/git --version" by called the function quite_system located in utils.rb. quiet_system just proxies to system in the same file which is where the hangup occurs. 

Interesting enough, when you reach the Process.wait, you can lookup and manually terminate the process via the PID and the install completes successfully.

**Takeaway**
No idea why this is occurring or if we should open a ticket with brew. Sorry for the wall of text but I've been working on this for awhile and am a bit fatigued. Figured I'd at least pass along what I've discovered over the last few days.

**Comment by aaronbruckner** [comment]
With some extra puts, the only log I see is "FORK". I haven't touched ruby so I'm not exactly sure what the line "yield if block_given?" is doing.

```
def self._system(cmd, *args)
    pid = fork do
      puts "FORK"
      yield if block_given?
      puts "1"
      args.collect!(&:to_s)
      puts "2"
      exec(cmd, *args) rescue nil
      puts "3"
      exit! 1 # never gets here unless exec failed
    end
    puts "Process ID: " + pid.to_s
    Process.wait(pid)
    puts "Process complete!"
    $?.success?
  end
```

**Comment by aaronbruckner** [comment]
Oh, that yield if block_given is executing code within quiet_system.

With some extra prints, the last line I see is "Reopen!" which means the trouble line is $stdout.reopen("/dev/null").  However the exec still doesn't complete if I comment out the stdout lines when it returns to the system function.

```
def quiet_system(cmd, *args)
  Homebrew._system(cmd, *args) do
    # Redirect output streams to `/dev/null` instead of closing as some programs
    # will fail to execute if they can't write to an open stream.
    puts "Reopen!"
    $stdout.reopen("/dev/null")
    puts "reopen1"
    $stderr.reopen("/dev/null")
    puts "reopen2"
  end
end
```

**Comment by denji** [comment]
I suguess the problem is in homebrew-core?

**Comment by aaronbruckner** [comment]
I'm not a betting man but it seems most likely. All this feel pretty far removed from the formula.

**Comment by julianhirt** [comment]
I just updated git to version 2.8.0 and it worked like a charm.

**Comment by aaronbruckner** [comment]
Same! Updated via brew to 2.8 and I am no longer having any issues.

**Comment by wa0x6e** [comment]
+1, updating to git 2.8 solved the issue

**Comment by brunnolou** [comment]
:+1:,  `brew install git` + new terminal window solved!
